pkg/discovery/dtofactory/volume_entity_dto_builder.go

```diff
@@ -0,0 +1,154 @@
+package dtofactory
+
+import (
+	"fmt"
+
+	api "k8s.io/api/core/v1"
+
+	"github.com/turbonomic/kubeturbo/pkg/discovery/dtofactory/property"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/stitching"
+	sdkbuilder "github.com/turbonomic/turbo-go-sdk/pkg/builder"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+
+	"github.com/golang/glog"
+)
+
+type volumeEntityDTOBuilder struct {
+	podVolumeMetrics []*repository.PodVolumeMetrics
+}
+
+func NewVolumeEntityDTOBuilder(podVolumeMetrics []*repository.PodVolumeMetrics) *volumeEntityDTOBuilder {
+	return &volumeEntityDTOBuilder{
+		podVolumeMetrics: podVolumeMetrics,
+	}
+}
+
+// Build entityDTOs based on the given volume to pod mappings.
+func (builder *volumeEntityDTOBuilder) BuildEntityDTOs(volToPodsMap map[*api.PersistentVolume][]repository.PodVolume) ([]*proto.EntityDTO, error) {
+	var result []*proto.EntityDTO
+
+	for vol, podVolumes := range volToPodsMap {
+		volID := string(vol.UID)
+		entityDTOBuilder := sdkbuilder.NewEntityDTOBuilder(proto.EntityDTO_VIRTUAL_VOLUME, volID)
+		displayName := vol.Name
+		entityDTOBuilder.DisplayName(displayName)
+
+		commoditiesSold, cap, err := builder.getVolumeCommoditiesSold(vol, podVolumes)
+		if err != nil {
+			glog.Errorf("Error creating commoditiesSold for volume %s: %s", displayName, err)
+		}
+
+		entityDTOBuilder.SellsCommodities(commoditiesSold)
+
+		var vols []*api.PersistentVolume
+		vols = append(vols, vol)
+		volStitchingMgr := stitching.NewVolumeStitchingManager()
+		err = volStitchingMgr.ProcessVolumes(vols)
+		if err != nil {
+			glog.Errorf("Error generating stitching metadata for volume %s: %s", displayName, err)
+		} else {
+			// reconciliation meta data
+			metaData, err := volStitchingMgr.GenerateReconciliationMetaData()
+			if err != nil {
+				glog.Errorf("Failed to build reconciliation metadata for volume %s: %s", displayName, err)
+			}
+			entityDTOBuilder = entityDTOBuilder.ReplacedBy(metaData)
+		}
+
+		// entities' properties.
+		properties := builder.getVolumeProperties(vol, volStitchingMgr)
+		entityDTOBuilder = entityDTOBuilder.WithProperties(properties)
+
+		entityDTOBuilder.WithPowerState(proto.EntityDTO_POWERED_ON)
+
+		// build entityDTO.
+		entityDto, err := entityDTOBuilder.Create()
+		if err != nil {
+			glog.Errorf("Failed to build Volume: %s entityDTO: %s", displayName, err)
+			continue
+		}
+
+		volCapacity := float32(cap)
+		isEphemaral := false
+		virtualVolumeData := &proto.EntityDTO_VirtualVolumeData{
+			StorageAmountCapacity: &volCapacity,
```

Comment:
Looks `StorageAmountCapacity` is deprecated in `EntityDTO_VirtualVolumeData`, and the comment of this field in `CommonDTO.pb.go` says:
> The field is deprecated because the value is now carried in the sold commodity

If  I understand correctly, we may not need to set volCapacity here, instead the StorageAmount commodity sold handles this capacity.

Reply:
I am already setting the capacity in the sold commodity. I can't remember it completely now, but I think I did encounter some issue on an older server if this was not set and thats probably the reason I did introduce it. Is there any harm in leaving it as is, even if the value is unused by the server?

Reply:
Actually I don't think there's any harm with this (it may be fine every from memory perspective). I was just wondering if we could avoid some redundancies since it's marked as deprecated. Still like you said, it could cause some issue on an older server. So looks It would be safer to keep this